package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -679,90 +679,136 @@ def _get_dh_pairs(self):
-        for group1, group2 in self.between_ags:
-
-            # Find donors in G1 and acceptors in G2
-            mask[
-                    np.logical_and(
-                        np.isin(donors.indices, group1.indices),
-                        np.isin(acceptors.indices, group2.indices)
-                    )
-            ] = True
-
-            # Find acceptors in G1 and donors in G2
-            mask[
-                np.logical_and(
-                    np.isin(acceptors.indices, group1.indices),
-                    np.isin(donors.indices, group2.indices)
-                )
-            ] = True
-
-        return mask
-
-
     def _prepare(self):
         self.results.hbonds = [[], [], [], [], [], []]
-
-
+ 
     def _single_frame(self):
-
         box = self._ts.dimensions
 
         # Update donor-hydrogen pairs if necessary
         if self.update_selections:
             self._donors, self._hydrogens = self._get_dh_pairs()
 
-        # find D and A within cutoff distance of one another
-        # min_cutoff = 1.0 as an atom cannot form a hydrogen bond with itself
-        d_a_indices, d_a_distances = capped_distance(
-            self._donors.positions,
-            self._acceptors.positions,
-            max_cutoff=self.d_a_cutoff,
-            min_cutoff=1.0,
-            box=box,
-            return_distances=True,
-        )
+        if self.between_ags is not None:
+            # map donor indices to respective hydrogen indices
+            donor_idx_to_H_indices = defaultdict(list)
+
+            for d, h in zip(self._donors, self._hydrogens):
+                donor_idx_to_H_indices[d.index].append(h.index)
+
+            tmp_hbond_candidates = [] 
+
+            for group1, group2 in self.between_ags:
+
+                tmp_donors_g1_mask = np.isin(self._donors.indices, group1.indices) 
+                donors_for_dist = self._donors[tmp_donors_g1_mask] 
+                tmp_acceptors_g2_mask = np.isin(self._acceptors.indices, group2.indices) 
+                acceptors_for_dist = self._acceptors[tmp_acceptors_g2_mask] 
+ 
+                d_a_indices_12, d_a_distances_12 = capped_distance(
+                    donors_for_dist.positions,
+                    acceptors_for_dist.positions,
+                    max_cutoff=self.d_a_cutoff,
+                    box=box,
+                    return_distances=True,
+                )
 
-        if np.size(d_a_indices) == 0:
-            warnings.warn(
-                "No hydrogen bonds were found given d-a cutoff of "
-                f"{self.d_a_cutoff} between Donor, {self.donors_sel}, and "
-                f"Acceptor, {self.acceptors_sel}."
+                if np.size(d_a_indices_12) == 0:
+                    warnings.warn(
+                        "No hydrogen bonds were found given d-a cutoff of "
+                        f"{self.d_a_cutoff} between Donor, {self.donors_sel}, and "
+                        f"Acceptor, {self.acceptors_sel}."
```

Comment:
this warning is no longer accurate - it should warn that no hydrogen bonds were found _between groups 1 and 2_ for the given cutoff and atom selections